### PR TITLE
[Merged by Bors] - feat: the single functors from the homotopy category

### PR DESCRIPTION
**Pull request by joelriou**
In this PR, we define the single functors from a preadditive category (with a zero object) to its homotopy category of cochain complexes and show they behave well with respect to shift functors.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
Judging by the comments in the code, I have marked this `WIP`.

**Comment by joelriou** [comment]
Now this is ready for review.

**Review by erdOne — COMMENTED** [review]
I think `simps` can generate some useful lemmas?

**Comment by joelriou** [comment]
> I think `simps` can generate some useful lemmas?

Actually, in most (future) situations, we do not want these definitions to be unfolded. These are almost like blackboxes which give an additional structure on these categories.

**Comment by erdOne** [comment]
Gotcha. Can you record this in the docstring?

**Comment by joelriou** [comment]
I have just added a comment.

File: Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+variable (C : Type u) [Category.{v} C] [Preadditive C] [HasZeroObject C]
+
+namespace CochainComplex
+
+open HomologicalComplex
+
+/-- The collection of all single functors `C ⥤ CochainComplex C ℤ` along with
+their compatibilites with shifts. (This definition has purposely no `simps`
+attribute, as the generated lemmas would be very useful.) -/
```
Comment (erdOne) [inline Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean:33]: I don't think this is what you meant?
Or maybe even `irreducible_def` if you are 100% sure.

Comment (joelriou) [reply Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean]: I have fixed the phrasing. (`irreducible_def` would be too much.)

File: Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    ext
+    simp [shiftFunctorAdd_eq, XIsoOfEq]
+
+instance (n : ℤ) : ((singleFunctors C).functor n).Additive := by
+  dsimp only [singleFunctors]
+  infer_instance
+
+/-- The single functor `C ⥤ CochainComplex C ℤ` which sends `X` to the complex
+consisting of `X` in degree `n : ℤ` and zero otherwise. -/
+noncomputable abbrev singleFunctor (n : ℤ) := (singleFunctors C).functor n
```
Comment (erdOne) [inline Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean:62]: So this is `defeq` to `single _ _ n` but this is the main API and not `single` right?
I think it is useful to document these things. (This is what I meant above)

Comment (joelriou) [reply Mathlib/Algebra/Homology/HomotopyCategory/SingleFunctors.lean:64]: Thanks for the suggestion.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**